### PR TITLE
Skipper undeploy attempts delete of other info when no package is found for stream

### PR DESCRIPTION
**Pull request by onobc**
Fixes GH-5373

### CODE REVIEW

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/stream/SkipperStreamDeployer.java [hunk 1]
```diff
@@ -463,15 +463,20 @@ private List<Template> createGenericTemplate() {
...
+        if (packageExists) {
             try {
                 this.skipperClient.delete(streamName, true);
             } catch (ReleaseNotFoundException e) {
                 logger.info("Release not found for {}. Deleting the package {}", streamName, streamName);
                 this.skipperClient.packageDelete(streamName);
             }
+            return;
         }
+        // The assumption is the package data is corrupted/missing so we attempt to delete all but the package info
```
Comment (onobc) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/stream/SkipperStreamDeployer.java:477]: We should ultimately provide a boolean flag that can be passed in to decide whether to do this. It is only useful in the case where a customer has "corrupted" data and we want to force a delete regardless of what is going on.

Let's discuss @markpollack @corneil 

If we want to add a flag this is out of scope for `2.11.0` (IMO).

File: spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/stream/SkipperStreamDeployerTests.java [hunk 2]
```diff
@@ -499,40 +501,67 @@ public void testStateOfDefinedUndeployedStream() {
...
 		SkipperClient skipperClient = mock(SkipperClient.class);
 		StreamDefinitionRepository streamDefinitionRepository = mock(StreamDefinitionRepository.class);
-
 		SkipperStreamDeployer skipperStreamDeployer = new SkipperStreamDeployer(skipperClient,
 				streamDefinitionRepository, appRegistryService, mock(ForkJoinPool.class), new DefaultStreamDefinitionService());
-
 		StreamDefinition streamDefinition = new StreamDefinition("foo", "foo|bar");
-
-		when(skipperClient.search(eq(streamDefinition.getName()), eq(false))).thenReturn(new ArrayList<>());
+		PackageMetadata packageMetadata = new PackageMetadata();
```
Comment (onobc) [inline spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/stream/SkipperStreamDeployerTests.java:511]: These tests are too mocked out - only really verifying that the impl I have added is in fact what gets called. Ultimately we need to do the following in a full integration test:

- Create stream `foo`
- Create stream `foobar`
- Delete foo package data (for real or mocked out)
- Delete stream `foo`
- Verify stream `foo` deleted
- Create stream `foo` 
- Verify it creates stream w/o barfing about "release already exists".

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/stream/SkipperStreamDeployer.java (hunk 1 above)
Comment (corneil) [reply spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/stream/SkipperStreamDeployer.java:477]: We probably don't want to fail a delete for something that doesn't exist. It is the same type of scenario like when multiple tasks are deleted from UI using forkJoin which runs mutiple deletes in parallel, some have children that get deleted when deleting parent and then when the child is tried and doesn't exists it fails. We don't need to fail if delete doesn't actually delete anything since another delete may have been done just before by another party.

File: spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/stream/SkipperStreamDeployerTests.java (hunk 2 above)
Comment (corneil) [reply spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/stream/SkipperStreamDeployerTests.java:511]: We need a test that launches rabbit, skipper and database using containers and then using streamservice to deploy tests and jdbctemplate to delete the entries. Then service to perform delete

Comment (onobc) [reply spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/stream/SkipperStreamDeployerTests.java:511]: Exactly.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/stream/SkipperStreamDeployer.java (hunk 1 above)
Comment (onobc) [reply spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/stream/SkipperStreamDeployer.java:477]: > We probably don't want to fail a delete for something that doesn't exist. 

Not sure where you are referring to. We are skipping the delete when the package does not exist which prevents the failing delete from happening. There are 2 changes above:

### Stop wildcard search from allowing streams w/o package metadata to be deleted
```java
        boolean packageExists = packages.stream().map(PackageMetadata::getName).anyMatch(s -> s.equalsIgnoreCase(streamName));
        if (packageExists) {
              // do same thing as before 'delete(name, true)'
        }
```
Before the above change, if package metadata was missing for "foo" (but there was a stream "foobar") then it would attempt the delete (including package data). The delete would fail, but not with the caught exception `ReleaseNotFoundException` and instead a `SkipperException` and hence it would bubble up to the caller and the stream def would not be deleted. Therefore the user was never able to delete the stream def from the UI.

The above is effectively an exact name (case insensitive) search so that delete will only work if the package data exists. This allows the user to delete the stream def and all "appears" fine.  It is not though, see next section.

### Allow stream to be re-created w/ same name
The 2nd part of the code change is to delete the stream when the package does not exist (excluding package metadata because we know it does not exist). 
```java
  logger.warn("Can not find package named '{}' - assuming package corrupted; attempting to delete release info", streamName);
        this.skipperClient.delete(streamName, false);
```
Note the `false` param. 

Without this change, the next issue is that when the user attempts to re-create the stream w/ same name (eg. "foo"), it fails w/ a "Release w/ name 'foo' already exists" because it did not remove the other data/state from Skipper as we skipped the delete during "undeploy".

Adding this other step clears deletes any other stream data and solves both of the cases. 

### My 2c
I think this is the wrong way to go. I think it would be better to provide a dedicated "cleanup stream" API that could be used in the case where a user has corrupted data (for whatever reason). The API would delete each piece of the stream in a safe manner and ensure the data is gone (all of it). This could be secured and under the "Misc" panel in the UI. Or shell access only??

I also think this is growing in scope and is not as important as getting `2.11.0` out.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/stream/SkipperStreamDeployer.java [hunk 3]
```diff
@@ -463,14 +463,18 @@ private List<Template> createGenericTemplate() {
...
+        if (packageExists) {
             try {
                 this.skipperClient.delete(streamName, true);
             } catch (ReleaseNotFoundException e) {
                 logger.info("Release not found for {}. Deleting the package {}", streamName, streamName);
                 this.skipperClient.packageDelete(streamName);
-            }
+            };
+        } else {
+            // Stream may have package data is corrupted/missing so we attempt to delete all but the package info
```
Comment (cppwfs) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/stream/SkipperStreamDeployer.java:476]: Nitpick: package data `that` is

Comment (onobc) [reply spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/stream/SkipperStreamDeployer.java]: Good catch - actually the comment is outdate altogether - yanked.
